#include "SelectDatabaseDialog.h"
#include "ui_SelectDatabaseDialog.h"

#include "EditDatabaseDialog.h"

#include <QMessageBox>
#include <QSettings>

namespace SC::UI::Dialogs {

SelectDatabaseDialog::SelectDatabaseDialog(QWidget *parent)
    : QDialog(parent)
    , ui(new Ui::SelectDatabaseDialog)
    , m_model(new QStandardItemModel(this))
{
    ui->setupUi(this);    

    setupModel();
    loadConnections();

    ui->tableViewDatabases->setModel(m_model);

    // üîπ –ü–æ–∫–∞–∑—É—î–º–æ –ª–∏—à–µ —Ä—è–¥–∫–∏
    ui->tableViewDatabases->setSelectionBehavior(QAbstractItemView::SelectRows);
    ui->tableViewDatabases->setSelectionMode(QAbstractItemView::SingleSelection);

    // üîπ –ü—Ä–∏–±–∏—Ä–∞—î–º–æ –∑–∞–≥–æ–ª–æ–≤–∫–∏
    ui->tableViewDatabases->horizontalHeader()->hide();
    ui->tableViewDatabases->verticalHeader()->hide();

    // üîπ –•–æ–≤–∞—î–º–æ –≤—Å—ñ –∫–æ–ª–æ–Ω–∫–∏, –∫—Ä—ñ–º –ø–µ—Ä—à–æ—ó
    for (int col = 1; col < m_model->columnCount(); ++col) {
        ui->tableViewDatabases->setColumnHidden(col, true);
    }

    // üîπ –ü–µ—Ä—à–∞ –∫–æ–ª–æ–Ω–∫–∞ –Ω–∞ –≤—Å—é —à–∏—Ä–∏–Ω—É
    ui->tableViewDatabases->horizontalHeader()
        ->setSectionResizeMode(0, QHeaderView::Stretch);

    // üîπ –ë–µ–∑ —Å—ñ—Ç–∫–∏ (–≤–∏–≥–ª—è–¥–∞—î —è–∫ —Å–ø–∏—Å–æ–∫)
    ui->tableViewDatabases->setShowGrid(false);

    // üîπ –û–ø—Ü—ñ–æ–Ω–∞–ª—å–Ω–æ: –±–µ–∑ —Ä–∞–º–∫–∏
    ui->tableViewDatabases->setFrameShape(QFrame::NoFrame);

    connect(ui->btnAdd,    &QPushButton::clicked, this, &SelectDatabaseDialog::addDatabase   );
    connect(ui->btnEdit,   &QPushButton::clicked, this, &SelectDatabaseDialog::editDatabase  );
    connect(ui->btnRemove, &QPushButton::clicked, this, &SelectDatabaseDialog::removeDatabase);
    connect(ui->btnSelect, &QPushButton::clicked, this, &SelectDatabaseDialog::selectDatabase);
    connect(ui->btnCancel, &QPushButton::clicked, this, &QDialog::reject                     );
}

SelectDatabaseDialog::~SelectDatabaseDialog()
{
    delete ui;
}

/* =========================
 *  Public
 * ========================= */

SC::Application::Database::DatabaseConnectionInfo SelectDatabaseDialog::selectedDatabase() const
{
    SC::Application::Database::DatabaseConnectionInfo info;

    if (m_selectedRow < 0)
        return info;

    info.name     = m_model->item(m_selectedRow, 0)->text();
    info.host     = m_model->item(m_selectedRow, 1)->text();
    info.port     = m_model->item(m_selectedRow, 2)->text().toInt();
    info.database = m_model->item(m_selectedRow, 3)->text();
    info.user     = m_model->item(m_selectedRow, 4)->text();

    return info;
}

/* =========================
 *  Slots
 * ========================= */

void SelectDatabaseDialog::addDatabase()
{

    EditDatabaseDialog dlg(this);

    if (dlg.exec() != QDialog::Accepted)
        return;

    const auto info = dlg.data();

    if (info.name.trimmed().isEmpty()) {
        QMessageBox::warning(this, tr("–ü–æ–º–∏–ª–∫–∞"),
                             tr("–ù–∞–∑–≤–∞ –ø—ñ–¥–∫–ª—é—á–µ–Ω–Ω—è –Ω–µ –º–æ–∂–µ –±—É—Ç–∏ –ø–æ—Ä–æ–∂–Ω—å–æ—é"));
        return;
    }

    QList<QStandardItem*> row;
    row << new QStandardItem(info.name    );
    row << new QStandardItem(info.host    );
    row << new QStandardItem(QString::number(info.port));
    row << new QStandardItem(info.database);
    row << new QStandardItem(info.user    );
    row << new QStandardItem(info.password);

    m_model->appendRow(row);
    saveConnections();
}

void SelectDatabaseDialog::editDatabase()
{

    QModelIndex index = ui->tableViewDatabases->currentIndex();
    if (!index.isValid()) {
        QMessageBox::warning(
            this,
            tr("–†–µ–¥–∞–≥—É–≤–∞–Ω–Ω—è –ø—ñ–¥–∫–ª—é—á–µ–Ω–Ω—è"),
            tr("–ë—É–¥—å –ª–∞—Å–∫–∞, –≤–∏–±–µ—Ä—ñ—Ç—å –ø—ñ–¥–∫–ª—é—á–µ–Ω–Ω—è –∑—ñ —Å–ø–∏—Å–∫—É.")
            );
        return;
    }

    const int row = index.row();
    QStandardItem* item = m_model->item(row, 0);
    if (!item) {
        return;
    }

    // 1. –î—ñ—Å—Ç–∞—î–º–æ DatabaseConnectionInfo –∑ –º–æ–¥–µ–ª—ñ
    SC::Application::Database::DatabaseConnectionInfo info;
    info.name     = m_model->item(row, 0)->text();
    info.host     = m_model->item(row, 1)->text();
    info.port     = m_model->item(row, 2)->text().toInt(); // –ö–æ–Ω–≤–µ—Ä—Ç—É—î–º–æ –Ω–∞–∑–∞–¥ —É —á–∏—Å–ª–æ
    info.database = m_model->item(row, 3)->text();
    info.user     = m_model->item(row, 4)->text();
    info.password = m_model->item(row, 5)->text();

    // 2. –î—ñ–∞–ª–æ–≥ —Ä–µ–¥–∞–≥—É–≤–∞–Ω–Ω—è
    EditDatabaseDialog dlg(info, this);
    if (dlg.exec() != QDialog::Accepted)
        return;

    // 3. –û—Ç—Ä–∏–º—É—î–º–æ –æ–Ω–æ–≤–ª–µ–Ω—ñ –¥–∞–Ω—ñ
    const auto updatedInfo = dlg.data();

    if (updatedInfo.name.trimmed().isEmpty()) {
        QMessageBox::warning(this, tr("–ü–æ–º–∏–ª–∫–∞"),
                             tr("–ù–∞–∑–≤–∞ –ø—ñ–¥–∫–ª—é—á–µ–Ω–Ω—è –Ω–µ –º–æ–∂–µ –±—É—Ç–∏ –ø–æ—Ä–æ–∂–Ω—å–æ—é"));
        return;
    }

    // 4. –û–Ω–æ–≤–ª—é—î–º–æ —Ç–µ–∫—Å—Ç —É –∫–æ–∂–Ω—ñ–π –∫–ª—ñ—Ç–∏–Ω—Ü—ñ –ø–æ—Ç–æ—á–Ω–æ–≥–æ —Ä—è–¥–∫–∞
    m_model->item(row, 0)->setText(updatedInfo.name);
    m_model->item(row, 1)->setText(updatedInfo.host);
    m_model->item(row, 2)->setText(QString::number(updatedInfo.port));
    m_model->item(row, 3)->setText(updatedInfo.database);
    m_model->item(row, 4)->setText(updatedInfo.user);
    m_model->item(row, 5)->setText(updatedInfo.password);

    // 5. –í–ê–ñ–õ–ò–í–û: –û–Ω–æ–≤–ª—é—î–º–æ –¥–∞–Ω—ñ –≤ UserRole, —â–æ–± –ø—Ä–∏ –Ω–∞—Å—Ç—É–ø–Ω–æ–º—É –Ω–∞—Ç–∏—Å–∫–∞–Ω–Ω—ñ
    // "–†–µ–¥–∞–≥—É–≤–∞—Ç–∏" –¥—ñ–∞–ª–æ–≥ –æ—Ç—Ä–∏–º–∞–≤ —Å–≤—ñ–∂—ñ –¥–∞–Ω—ñ, –∞ –Ω–µ —Å—Ç–∞—Ä—ñ
    m_model->item(row, 0)->setData(QVariant::fromValue(updatedInfo), Qt::UserRole);

    // 6. UX: –∑–∞–ª–∏—à–∞—î–º–æ –≤–∏–¥—ñ–ª–µ–Ω–Ω—è
    ui->tableViewDatabases->setCurrentIndex(
        m_model->index(row, 0)
        );

    // 7. –ü—ñ—Å–ª—è –æ–Ω–æ–≤–ª–µ–Ω–Ω—è –º–æ–¥–µ–ª—ñ –∑–∞–∑–≤–∏—á–∞–π –ø–æ—Ç—Ä—ñ–±–Ω–æ –∑–±–µ—Ä–µ–≥—Ç–∏ –∑–º—ñ–Ω–∏ —É QSettings
    saveConnections();
}

void SelectDatabaseDialog::removeDatabase()
{
    auto index = ui->tableViewDatabases->currentIndex();
    if (!index.isValid())
        return;

    m_model->removeRow(index.row());
    saveConnections();
}

void SelectDatabaseDialog::selectDatabase()
{
    auto index = ui->tableViewDatabases->currentIndex();
    if (!index.isValid()) {
        QMessageBox::warning(this, tr("–ü–æ–º–∏–ª–∫–∞"),
                             tr("–û–±–µ—Ä—ñ—Ç—å –±–∞–∑—É –¥–∞–Ω–∏—Ö"));
        return;
    }

    m_selectedRow = index.row();
    accept();
}

/* =========================
 *  Private
 * ========================= */

void SelectDatabaseDialog::setupModel()
{
    m_model->setColumnCount(5);
    m_model->setHorizontalHeaderLabels({
        tr("–ù–∞–∑–≤–∞"     ),
        tr("Host"      ),
        tr("Port"      ),
        tr("–ë–∞–∑–∞"      ),
        tr("–ö–æ—Ä–∏—Å—Ç—É–≤–∞—á"),
        tr("–ü–∞—Ä–æ–ª—å"    )
    });
}

void SelectDatabaseDialog::loadConnections()
{
    m_model->removeRows(0, m_model->rowCount()); // –û—á–∏—â–µ–Ω–Ω—è —Å—Ç–∞—Ä–∏—Ö –¥–∞–Ω–∏—Ö
    QSettings settings;

    int size = settings.beginReadArray("db_connections");
    for (int i = 0; i < size; ++i) {
        settings.setArrayIndex(i);

        // –°—Ç–≤–æ—Ä—é—î–º–æ —Å–ø–∏—Å–æ–∫ –µ–ª–µ–º–µ–Ω—Ç—ñ–≤ –¥–ª—è –æ–¥–Ω–æ–≥–æ —Ä—è–¥–∫–∞
        QList<QStandardItem*> row {
            new QStandardItem(settings.value("name"    ).toString()),
            new QStandardItem(settings.value("host"    ).toString()),
            new QStandardItem(settings.value("port"    ).toString()),
            new QStandardItem(settings.value("database").toString()),
            new QStandardItem(settings.value("user"    ).toString()),
            new QStandardItem(settings.value("password").toString())
        };

        m_model->appendRow(row); // –¢–µ–ø–µ—Ä –º–æ–¥–µ–ª—å –≤—ñ–¥–ø–æ–≤—ñ–¥–∞—î –∑–∞ –ø–∞–º'—è—Ç—å
    }
    settings.endArray();
}

void SelectDatabaseDialog::saveConnections()
{
    QSettings settings;
    settings.beginWriteArray("db_connections");

    for (int row = 0; row < m_model->rowCount(); ++row) {
        settings.setArrayIndex(row);
        settings.setValue("name",     m_model->item(row, 0)->text());
        settings.setValue("host",     m_model->item(row, 1)->text());
        settings.setValue("port",     m_model->item(row, 2)->text());
        settings.setValue("database", m_model->item(row, 3)->text());
        settings.setValue("user",     m_model->item(row, 4)->text());
        settings.setValue("password", m_model->item(row, 5)->text());
    }

    settings.endArray();
}

} // namespace SC::UI::Dialogs
